### PR TITLE
[bitnami/aspnet-corel] New Chart

### PR DESCRIPTION
**Pull request by shankarmohannamakkal**
**Description of the change**

New chart for deploying ASP.NET Core applications.

**Checklist**

- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [X] Variables are documented in the README.md
- [X] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)

### CODE REVIEW

**Comment by carrodher** [comment]
Hi, thank you very much for your contribution! 🎉 
I was taking a quick look and at first glance, it seems at this moment it is not functional due to the image.
Are you thinking on add a working image for the chart? One of the few requirements to publish charts in the Bitnami repository is to use Bitnami images for the containers.

We already have an aspnet-core image, see the [GitHub](https://github.com/bitnami/bitnami-docker-aspnet-core) and [DockerHub](https://hub.docker.com/repository/docker/bitnami/aspnet-core) repositories. I just tried the chart using this image but it seems it doesn't work out of the box:
```console
$ helm install dotnet . --set image.registry=docker.io --set image.repository=bitnami/aspnet-core --set image.tag=3.1.5-debian-10-r7
NAME: dotnet
LAST DEPLOYED: Wed Jun 17 16:41:37 2020
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
** Please be patient while the chart is being deployed **

  NOTE: It may take a few minutes for the IP to be available.
        Watch the status with: 'kubectl get svc --namespace default -w dotnet-dotnet-core-kestrel'

  export SERVICE_IP=$(kubectl get svc --namespace default dotnet-dotnet-core-kestrel --template "{{ range (index .status.loadBalancer.ingress 0) }}{{.}}{{ end }}")
  echo "NGINX URL: http://$SERVICE_IP/"

$ kubectl get pods
NAME                                           READY   STATUS             RESTARTS   AGE
dotnet-dotnet-core-kestrel-6b9b75bfcf-q92gq    0/1     CrashLoopBackOff   6          7m57s

$ kubectl describe pod dotnet-dotnet-core-kestrel-6b9b75bfcf-q92gq
Name:         dotnet-dotnet-core-kestrel-6b9b75bfcf-q92gq
Namespace:    default
Priority:     0
Node:         gke-dev-default-pool-ab651c88-nmbb/10.130.47.237
Start Time:   Wed, 17 Jun 2020 16:34:21 +0000
Labels:       app.kubernetes.io/instance=dotnet
              app.kubernetes.io/managed-by=Helm
              app.kubernetes.io/name=dotnet-core-kestrel
              helm.sh/chart=dotnet-core-kestrel-0.1.0
              pod-template-hash=6b9b75bfcf
Annotations:  kubernetes.io/limit-ranger: LimitRanger plugin set: cpu request for container dotnet-core-kestrel
              kubernetes.io/psp: 60-mayroot
Status:       Running
IP:           10.30.11.179
IPs:
  IP:           10.30.11.179
Controlled By:  ReplicaSet/dotnet-dotnet-core-kestrel-6b9b75bfcf
Containers:
  dotnet-core-kestrel:
    Container ID:   docker://59927972eec68e29b0d939762d36fbb1cc2e097e732b37ce8563399d8dcae2ba
    Image:          docker.io/bitnami/aspnet-core:3.1.5-debian-10-r7
    Image ID:       docker-pullable://bitnami/aspnet-core@sha256:5f82b5589713479ec3add2baf90d4496c3e9c4ed2604023fe9bd7d3e0be38eaf
    Port:           8080/TCP
    Host Port:      0/TCP
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 17 Jun 2020 16:40:19 +0000
      Finished:     Wed, 17 Jun 2020 16:40:19 +0000
    Ready:          False
    Restart Count:  6
    Requests:
      cpu:        100m
    Environment:  <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-gd7xn (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  default-token-gd7xn:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-gd7xn
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason     Age                     From                                         Message
  ----     ------     ----                    ----                                         -------
  Normal   Scheduled  8m16s                   default-scheduler                            Successfully assigned default/dotnet-dotnet-core-kestrel-6b9b75bfcf-q92gq to gke-dev-default-pool-ab651c88-nmbb
  Normal   Pulled     6m38s (x5 over 8m15s)   kubelet, gke-dev-default-pool-ab651c88-nmbb  Container image "docker.io/bitnami/aspnet-core:3.1.5-debian-10-r7" already present on machine
  Normal   Created    6m38s (x5 over 8m15s)   kubelet, gke-dev-default-pool-ab651c88-nmbb  Created container dotnet-core-kestrel
  Normal   Started    6m38s (x5 over 8m15s)   kubelet, gke-dev-default-pool-ab651c88-nmbb  Started container dotnet-core-kestrel
  Warning  BackOff    3m13s (x26 over 8m13s)  kubelet, gke-dev-default-pool-ab651c88-nmbb  Back-off restarting failed container
```

Do you think it is possible that the chart works out of the box with this image? We're happy to help to create a functional chart and integrate it into the automated Bitnami CI/CD pipeline.

**Comment by shankarmohannamakkal** [comment]
Thanks @carrodher 

I have tested this chart with my private azure container registry asp.net core api image and it's working. 

I am not sure about the bitnami asp.net core images. Please help needful here.

**Comment by carrodher** [comment]
Hi, we are happy to help to make this chart works with the Bitnami image that you can find in [GitHub](https://github.com/bitnami/bitnami-docker-aspnet-core) and [DockerHub](https://hub.docker.com/repository/docker/bitnami/aspnet-core) repositories. Please, feel free to reach us regarding any issue you have deploying the chart with this image. If you think there is something we should implement in the image on our side, please, don't hesitate to let us know.

**Comment by shankarmohannamakkal** [comment]
> Hi, we are happy to help to make this chart works with the Bitnami image that you can find in [GitHub](https://github.com/bitnami/bitnami-docker-aspnet-core) and [DockerHub](https://hub.docker.com/repository/docker/bitnami/aspnet-core) repositories. Please, feel free to reach us regarding any issue you have deploying the chart with this image. If you think there is something we should implement in the image on our side, please, don't hesitate to let us know.

Thanks @carrodher.

I have uploaded a new docker image and updated the same in the chart values. I have tested this chart and its working without any change.

The docker image code is available in the git repo - https://github.com/shankarmohannamakkal/SampleAppplication

Please help to proceed with this....

**Comment by juan131** [comment]
Hi @shankarmohannamakkal 

I cloned your repo and I applied the changes below:

```diff
$ git diff
diff --git a/Dockerfile b/Dockerfile
index 7a317f6..9f2a6fc 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -13,7 +13,6 @@ WORKDIR /source/sampleapp
 RUN dotnet publish -c release -o /app --no-restore

 # final stage/image
-FROM mcr.microsoft.com/dotnet/core/aspnet:3.1
-WORKDIR /app
-COPY --from=build /app ./
-ENTRYPOINT ["dotnet", "sampleapp.dll"]
+FROM bitnami/aspnet-core:3.1-debian-10
+COPY --from=build /app /app
+ENV ASPNETCORE_URLS=http://+:80
+CMD ["dotnet", "sampleapp.dll"]
```

Then I run:
```
$ docker build . -t aspnetapp:1.0
$ docker run -it --rm -p 5000:80 --name aspnetcore_sample aspnetapp:1.0
$ curl http://localhost:5000/hello
Hello!
```

As you can see it's working using the Bitnam image as base image. We'll have to do sth very similar and create a Bitnami container for this chart.

cc @beltran-rubo @javsalgar

**Comment by shankarmohannamakkal** [comment]
> We'll have to do sth very similar and create a Bitnami container for this chart.

Thanks @juan131 for verify this.

Please help me to create a similar Bitnami container for this chart.

**Comment by juan131** [comment]
Hi @shankarmohannamakkal 

After discussing it with the team, we decided the best thing we can do is creating a "generic" ASP.NET Core chart similar to other framework charts we have such as [MXNet](https://github.com/bitnami/charts/tree/master/bitnami/mxnet), instead of having a specific one for Kestrel web server.

The idea is to have a generic solution for every ASP.NET Core app, so you can replace the default image [Bitnami ASP.NET Core](https://github.com/bitnami/bitnami-docker-aspnet-core) with your own one including your app's code (e.g the one you just created: `shankarmohan/aspnetapp`).

Do you mind if I contribute to your PR adding the needed changes to transform it into a generic solution? I would like to reuse as much as possible of the great work you did.

**Comment by shankarmohannamakkal** [comment]
> app

Sure, i like the idea of generic solution for every ASP.NET Core app.

Please go ahead and contribute to my PR adding the needed changes to transform it into a generic solution.

Thanks.

**Comment by juan131** [comment]
Great! Expect some commits from me during the next couple of days.

**Comment by juan131** [comment]
@shankarmohannamakkal I just push my changes converting the chart in a "generic" ASP.NET Core chart. Please give it a try when you have a chance.

@javsalgar @miguelaeh please review the PR when you have a chance

File: bitnami/aspnet-core/values.yaml [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+    ##
+    image:
+      registry: docker.io
+      repository: bitnami/git
+      tag: 2.27.0-debian-10-r32
+      ## Specify a imagePullPolicy
+      ## Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
+      ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
+      ##
+      pullPolicy: IfNotPresent
```
Comment (javsalgar) [inline bitnami/aspnet-core/values.yaml:58]: The value used in the yaml is `appFromExternalRepo.clone.pullPolicy`

File: bitnami/aspnet-core/values.yaml [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    ##
+    image:
+      registry: docker.io
+      repository: bitnami/dotnet-sdk
+      tag: 3.1.301-debian-10-r29
+      ## Specify a imagePullPolicy
+      ## Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
+      ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
+      ##
+      pullPolicy: IfNotPresent
```
Comment (javsalgar) [inline bitnami/aspnet-core/values.yaml:83]: Value used is `appFromExternalRepo.publish.pullPolicy`

File: bitnami/aspnet-core/values.yaml [hunk 3]
```diff
@@ -0,0 +1,379 @@
+## Global Docker image parameters
+## Please, note that this will override the image parameters, including dependencies, configured to use the global value
+## Current available global Docker image parameters: imageRegistry and imagePullSecrets
+##
+# global:
+#   imageRegistry: myRegistryName
+#   imagePullSecrets:
+#     - myRegistryKeySecretName
+
```
Comment (javsalgar) [inline bitnami/aspnet-core/values.yaml:9]: `priorityClassName` is not here

File: bitnami/aspnet-core/README.md [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+> Note: you can append any custom flag for the "dotnet publish" command setting the `appFromExternalRepo.publish.extraFlags` parameter.
+
+For example, you can deploy a sample [Kestrel server](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel) using the parameters below:
+
+```console
+appFromExternalRepo.enabled=true
+appFromExternalRepo.clone.repository=https://github.com/dotnet/AspNetCore.Docs.git
+appFromExternalRepo.clone.branch=master
+appFromExternalRepo.publish.aspnetcore/fundamentals/servers/kestrel/samples/3.x/KestrelSample
+appFromExternalRepo.startCommand[0]=dotnet
```
Comment (javsalgar) [inline bitnami/aspnet-core/README.md:227]: startCommand is not in the parameter table

File: bitnami/aspnet-core/README.md [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+| `ingress.secrets[0].certificate`        | TLS Secret Certificate                                              | `nil`                                                   |
+| `ingress.secrets[0].key`                | TLS Secret Key                                                      | `nil`                                                   |
+
+### RBAC parameters
+
+| Parameter                               | Description                                                         | Default                                                 |
+|-----------------------------------------|---------------------------------------------------------------------|---------------------------------------------------------|
+| `serviceAccount.create`                 | Enable the creation of a ServiceAccount for ASP.NET Core pods       | `true`                                                  |
+| `serviceAccount.name`                   | Name of the created serviceAccount                                  | Generated using the `aspnet-core.fullname` template     |
```
Comment (javsalgar) [inline bitnami/aspnet-core/README.md:165]: serviceAccount.annotations missing

File: bitnami/aspnet-core/README.md [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+| `service.loadBalancerIP`                | loadBalancerIP if service type is `LoadBalancer`                    | `nil`                                                   |
+| `service.loadBalancerSourceRanges`      | Address that are allowed when service is LoadBalancer               | `[]`                                                    |
+| `service.annotations`                   | Annotations for ASP.NET Core service                                | `{}`                                                    |
+| `ingress.enabled`                       | Enable ingress controller resource                                  | `false`                                                 |
+| `ingress.certManager`                   | Add annotations for cert-manager                                    | `false`                                                 |
+| `ingress.hostname`                      | Default host for the ingress resource                               | `aspnet-core.local`                                     |
```
Comment (javsalgar) [inline bitnami/aspnet-core/README.md:150]: ingress.tls missing

File: bitnami/aspnet-core/templates/_helpers.tpl [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+If release name contains chart name it will be used as a full name.
+*/}}
+{{- define "aspnet-core.fullname" -}}
+{{- include "common.names.fullname" . -}}
+{{- end }}
+
+{{/*
+Create chart name and version as used by the chart label.
+*/}}
+{{- define "aspnet-core.chart" -}}
```
Comment (javsalgar) [inline bitnami/aspnet-core/templates/_helpers.tpl:21]: Not used anywhere

File: bitnami/aspnet-core/templates/_helpers.tpl [hunk 8]
```diff
@@ -0,0 +1,85 @@
+{{/* vim: set filetype=mustache: */}}
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "aspnet-core.name" -}}
```
Comment (javsalgar) [inline bitnami/aspnet-core/templates/_helpers.tpl:5]: Not used anywhere

File: bitnami/aspnet-core/templates/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    rollingUpdate: null
+    {{- end }}
+  template:
+    metadata:
+      {{- if .Values.podAnnotations }}
+      annotations: {{- include "common.tplvalues.render" (dict "value" .Values.podAnnotations "context" $) | nindent 8 }}
+      {{- end }}
+      labels: {{- include "common.labels.standard" . | nindent 8 }}
+    spec:
+      {{- include "aspnet-core.imagePullSecrets" . | indent 6 }}
```
Comment (javsalgar) [inline bitnami/aspnet-core/templates/deployment.yaml:22]: nindent

File: bitnami/aspnet-core/templates/deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+      {{- end }}
+      {{- if .Values.podSecurityContext.enabled }}
+      securityContext:
+        fsGroup: {{ .Values.podSecurityContext.fsGroup }}
+        {{- if .Values.podSecurityContext.sysctls }}
+        sysctls:
+          {{- toYaml .Values.podSecurityContext.sysctls | nindent 10 }}
+        {{- end }}
+      {{- end }}
+      {{- if or .Values.appFromExternalRepo.enabled .Values.initContainers }}
```
Comment (javsalgar) [inline bitnami/aspnet-core/templates/deployment.yaml:44]: I wouldn't add this if, it adds noise to the code and the worst that would happen is an empty initContainers section

File: bitnami/aspnet-core/templates/deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,170 @@
+apiVersion: {{ include "common.capabilities.deployment.apiVersion" . }}
```
Comment (javsalgar) [inline bitnami/aspnet-core/templates/deployment.yaml:1]: If you have time, there are other values that we are including

- commonLabels
- commonAnnotations
- lifecycleHooks

Comment (javsalgar) [reply bitnami/aspnet-core/templates/deployment.yaml:1]: Example: harbor

**Comment by javsalgar** [comment]
I would also add extraDeploy, especially in a chart like this

File: bitnami/aspnet-core/templates/deployment.yaml (hunk 10 above)
Comment (juan131) [reply bitnami/aspnet-core/templates/deployment.yaml:44]: I'm very "picky" about generating "clean" YAMLs after running "helm template"

File: bitnami/aspnet-core/values.yaml [hunk 12]
```diff
@@ -38,8 +38,16 @@ image:
...
+##
+commonLabels: {}
+
+## Add annotations to all the deployed resources
+##
+commonAnnotations: {}
+
 ## Enable to download/build ASP.NET Core app from external git repository.
-## Disable it if your docker image already includes your application
+## Do not enabled it if your docker image already includes your application
```
Comment (javsalgar) [inline bitnami/aspnet-core/values.yaml:50]: enable

**Comment by shankarmohannamakkal** [comment]
> @shankarmohannamakkal I just push my changes converting the chart in a "generic" ASP.NET Core chart. Please give it a try when you have a chance.
> 
> @javsalgar @miguelaeh please review the PR when you have a chance

Thanks for the great support.

**Comment by juan131** [comment]
Thank you for the great contribution!! The chart was just published at https://charts.bitnami.com/bitnami

**Pull request merged**